[iOS] Stop using `-[NSURL _title]` and `-[NSURL _setTitle:]` when possible

whsieh opened the pull request:
#### 85f7fd9f6b7de860eac958c608b33b6a076a31b1
<pre>
[iOS] Stop using `-[NSURL _title]` and `-[NSURL _setTitle:]` when possible
<a href="https://bugs.webkit.org/show_bug.cgi?id=262471">https://bugs.webkit.org/show_bug.cgi?id=262471</a>
rdar://115803976

Reviewed by Tim Horton and Abrar Rahman Protyasha.

Remove uses of both `-[NSURL _setTitle:]` and `-[NSURL _title]`, in the case where
`HAVE(LINK_PRESENTATION_CHANGE_FOR_RADAR_115801517)` is set. Instead, take advantange of the fact
that LinkPresentation&apos;s `LPLinkMetadata` now sets `NSURL`&apos;s `_title` when calling `-setTitle:`,
after the changes in rdar://115801517. Similarly, LinkPresentation also returns the `NSURL`&apos;s
`_title` property in the case where no title is explicitly set, so we can simply set an NSURL and
then use `-[LPLinkMetadata title]` in place of calling `-_title` directly.

I verified that the following API tests (which exercise WebKit&apos;s ability to set and consume NSURL&apos;s
`_title`) continue to pass after this refactoring:

• ActionSheetTests.CopyImageElementWithHREFAndTitle
• ActionSheetTests.CopyImageElementWithHREF
• DragAndDropTests.LinkToInput
• DragAndDropTests.ExternalSourceTitledNSURL

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/cocoa/LinkPresentationSoftLink.h: Copied from Source/WebKit/Platform/spi/Cocoa/LinkPresentationSPI.h.
* Source/WebCore/PAL/pal/cocoa/LinkPresentationSoftLink.mm: Copied from Source/WebKit/Platform/spi/Cocoa/LinkPresentationSPI.h.

Add a LinkPresentation soft linking header, with `LPLinkMetadata` and `LPMetadataProvider`.

* Source/WebCore/PAL/pal/spi/cocoa/LinkPresentationSPI.h: Copied from Source/WebKit/Platform/spi/Cocoa/LinkPresentationSPI.h.

Move this SPI header from WebKit into WebCore.

* Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:

Remove now-unnecessary SPI declarations for `-_title` and `-_setTitle:`.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/NSURLUtilities.h: Copied from Source/WebKit/Platform/spi/Cocoa/LinkPresentationSPI.h.
* Source/WebCore/platform/cocoa/NSURLUtilities.mm: Renamed from Source/WebKit/Platform/spi/Cocoa/LinkPresentationSPI.h.
(-[NSURL _web_setTitle:]):
(-[NSURL _web_title]):

Add category helper methods to call `_title` and `_setTitle:` on an NSURL.

* Source/WebCore/platform/graphics/avfoundation/MediaSelectionGroupAVFObjC.h:
* Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
(WebCore::PlatformPasteboard::write):
(WebCore::PlatformPasteboard::readURL const):

Switch from using `-_setTitle:` and `-_title` to `-_web_setTitle:` and `-_web_title`, respectively.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/Cocoa/WKShareSheet.mm:

Adopt the new category helpers here as well. Note that in the case where we&apos;re adding a
`WKShareSheetURLItemProvider`, we can just pass in a `title` and `NSURL`, since LinkPresentation
now sets the `NSURL`&apos;s title after rdar://115801517.

(-[WKShareSheetFileItemProvider initWithURL:]):
(-[WKShareSheetURLItemProvider initWithURL:title:]):
(-[WKShareSheet presentWithParameters:inRect:completionHandler:]):
(-[WKShareSheetURLItemProvider initWithURL:]): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/268744@main">https://commits.webkit.org/268744@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d02db4565107279fab44fe05b45e16e946fadac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20540 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20964 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21613 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22436 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19160 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20774 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24219 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21141 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20760 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20615 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17863 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23294 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17782 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18668 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24945 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18860 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18844 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22883 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19428 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16508 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18643 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4933 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22982 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19247 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/PAL/pal/cocoa/LinkPresentationSoftLink.mm:
```diff
@@ -0,0 +1,38 @@
+
+#import "config.h"
+
+#if !PLATFORM(WATCHOS)
```
All of these things are available on watchOS now, I think you're propagating stale guards :)

whsieh replied:
Good point! I'll see if I can make this compile on non-internal watchOS SDKs.

aprotyas reviewed (approved):
LGTM. Same remark about the watchOS guard as Tim

whsieh committed:
[iOS] Stop using `-[NSURL _title]` and `-[NSURL _setTitle:]` when possible
https://bugs.webkit.org/show_bug.cgi?id=262471
rdar://115803976

Reviewed by Tim Horton and Abrar Rahman Protyasha.

Remove uses of both `-[NSURL _setTitle:]` and `-[NSURL _title]`, in the case where
`HAVE(LINK_PRESENTATION_CHANGE_FOR_RADAR_115801517)` is set. Instead, take advantange of the fact
that LinkPresentation's `LPLinkMetadata` now sets `NSURL`'s `_title` when calling `-setTitle:`,
after the changes in rdar://115801517. Similarly, LinkPresentation also returns the `NSURL`'s
`_title` property in the case where no title is explicitly set, so we can simply set an NSURL and
then use `-[LPLinkMetadata title]` in place of calling `-_title` directly.

I verified that the following API tests (which exercise WebKit's ability to set and consume NSURL's
`_title`) continue to pass after this refactoring:

• ActionSheetTests.CopyImageElementWithHREFAndTitle
• ActionSheetTests.CopyImageElementWithHREF
• DragAndDropTests.LinkToInput
• DragAndDropTests.ExternalSourceTitledNSURL

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/cocoa/LinkPresentationSoftLink.h: Copied from Source/WebKit/Platform/spi/Cocoa/LinkPresentationSPI.h.
* Source/WebCore/PAL/pal/cocoa/LinkPresentationSoftLink.mm: Copied from Source/WebKit/Platform/spi/Cocoa/LinkPresentationSPI.h.

Add a LinkPresentation soft linking header, with `LPLinkMetadata` and `LPMetadataProvider`.

* Source/WebCore/PAL/pal/spi/cocoa/LinkPresentationSPI.h: Copied from Source/WebKit/Platform/spi/Cocoa/LinkPresentationSPI.h.

Move this SPI header from WebKit into WebCore.

* Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:

Remove now-unnecessary SPI declarations for `-_title` and `-_setTitle:`.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/NSURLUtilities.h: Copied from Source/WebKit/Platform/spi/Cocoa/LinkPresentationSPI.h.
* Source/WebCore/platform/cocoa/NSURLUtilities.mm: Renamed from Source/WebKit/Platform/spi/Cocoa/LinkPresentationSPI.h.
(-[NSURL _web_setTitle:]):
(-[NSURL _web_title]):

Add category helper methods to call `_title` and `_setTitle:` on an NSURL.

* Source/WebCore/platform/graphics/avfoundation/MediaSelectionGroupAVFObjC.h:
* Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
(WebCore::PlatformPasteboard::write):
(WebCore::PlatformPasteboard::readURL const):

Switch from using `-_setTitle:` and `-_title` to `-_web_setTitle:` and `-_web_title`, respectively.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/Cocoa/WKShareSheet.mm:

Adopt the new category helpers here as well. Note that in the case where we're adding a
`WKShareSheetURLItemProvider`, we can just pass in a `title` and `NSURL`, since LinkPresentation
now sets the `NSURL`'s title after rdar://115801517.

(-[WKShareSheetFileItemProvider initWithURL:]):
(-[WKShareSheetURLItemProvider initWithURL:title:]):
(-[WKShareSheet presentWithParameters:inRect:completionHandler:]):
(-[WKShareSheetURLItemProvider initWithURL:]): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/268744@main
